File: capi/controller/internal/resources/cnpg.go
package resources

import (
	"fmt"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

func CNPGOperator(context Context, manifest []byte, taggedImage, exactImage string) ([]*unstructured.Unstructured, error) {
	objects, err := DecodeManifest(manifest)
	if err != nil {
		return nil, err
	}
	counts := map[string]int{}
	for _, object := range objects {
		replaceStrings(object.Object, map[string]string{taggedImage: exactImage}, counts)
		MarkTenantObject(context, object, "cnpg-operator")
	}
	if counts[taggedImage] != 2 {
		return nil, fmt.Errorf("unexpected CNPG operator image count")
	}
	return objects, nil
}

func CNPGObjects(context Context, storageClass, postgresImage string) []*unstructured.Unstructured {
	clusterName := "capi-postgres"
	values := []*unstructured.Unstructured{
		{Object: map[string]any{"apiVersion": "v1", "kind": "Namespace", "metadata": map[string]any{"name": "database"}}},
	}
	for ordinal := int32(1); ordinal <= context.Spec.DatabaseCount; ordinal++ {
		values = append(values, &unstructured.Unstructured{Object: map[string]any{
			"apiVersion": "v1", "kind": "PersistentVolume",
			"metadata": map[string]any{"name": fmt.Sprintf("%s-pv-%d", clusterName, ordinal)},
			"spec": map[string]any{
				"capacity": map[string]any{"storage": "1Gi"}, "accessModes": []any{"ReadWriteOnce"},
				"persistentVolumeReclaimPolicy": "Retain", "storageClassName": storageClass,
				"claimRef": map[string]any{"namespace": "database", "name": fmt.Sprintf("%s-%d", clusterName, ordinal)},
				"hostPath": map[string]any{
					"path": fmt.Sprintf("%s/volumes/cnpg/%d", context.Inputs.StorageContainerPath, ordinal),
					"type": "DirectoryOrCreate",
				},
			},
		}})
	}
	antiAffinity := "required"
	if context.Spec.DatabaseCount > context.Spec.Workers {
		antiAffinity = "preferred"
	}
	values = append(values, &unstructured.Unstructured{Object: map[string]any{
		"apiVersion": "postgresql.cnpg.io/v1", "kind": "Cluster",
		"metadata": map[string]any{"name": clusterName, "namespace": "database"},
		"spec": map[string]any{
			"instances": int64(context.Spec.DatabaseCount), "imageName": postgresImage,
			"affinity": map[string]any{
				"enablePodAntiAffinity": true, "podAntiAffinityType": antiAffinity, "topologyKey": "kubernetes.io/hostname",
			},
			"bootstrap": map[string]any{"initdb": map[string]any{"database": "app", "owner": "app"}},
			"storage":   map[string]any{"size": "1Gi", "storageClass": storageClass},
			"resources": map[string]any{
				"requests": map[string]any{"cpu": "100m", "memory": "256Mi"},
				"limits":   map[string]any{"cpu": "1", "memory": "1Gi"},
			},
		},
	}})
	for _, value := range values {
		MarkTenantObject(context, value, "cnpg")
	}
	return values
}

File: capi/controller/src/resources/cnpg.rs
use std::collections::BTreeMap;

use k8s_openapi::{
    api::{
        core::v1::{
            HostPathVolumeSource, Namespace, ObjectReference, PersistentVolume,
            PersistentVolumeSpec,
        },
        storage::v1::StorageClass,
    },
    apimachinery::pkg::api::resource::Quantity,
};
use kube::core::DynamicObject;
use serde_json::json;

use super::{
    BuildError, Context, decode_manifest, manifest::replace_object_strings, mark_tenant_object,
    to_dynamic,
};

pub fn storage_class(context: &Context<'_>, name: &str) -> StorageClass {
    StorageClass {
        metadata: context.metadata(name, "", "storage"),
        provisioner: "kubernetes.io/no-provisioner".into(),
        volume_binding_mode: Some("Immediate".into()),
        reclaim_policy: Some("Retain".into()),
        ..Default::default()
    }
}

pub fn cnpg_operator(
    context: &Context<'_>,
    manifest: &[u8],
    tagged_image: &str,
    exact_image: &str,
) -> Result<Vec<DynamicObject>, BuildError> {
    let mut objects = decode_manifest(manifest)?;
    let mut counts = BTreeMap::new();
    for object in &mut objects {
        replace_object_strings(
            object,
            &BTreeMap::from([(tagged_image, exact_image)]),
            &mut counts,
        )?;
        mark_tenant_object(context, object, "cnpg-operator");
    }
    if counts.get(tagged_image) != Some(&2) {
        return Err(BuildError::Manifest(
            "unexpected CNPG operator image count".into(),
        ));
    }
    Ok(objects)
}

pub fn database_namespace(context: &Context<'_>) -> Namespace {
    Namespace {
        metadata: context.metadata("database", "", "cnpg"),
        ..Default::default()
    }
}

pub fn persistent_volumes(context: &Context<'_>, storage_class: &str) -> Vec<PersistentVolume> {
    (1..=context.spec.databases)
        .map(|ordinal| PersistentVolume {
            metadata: context.metadata(&format!("capi-postgres-pv-{ordinal}"), "", "cnpg"),
            spec: Some(PersistentVolumeSpec {
                capacity: Some(BTreeMap::from([("storage".into(), Quantity("1Gi".into()))])),
                access_modes: Some(vec!["ReadWriteOnce".into()]),
                persistent_volume_reclaim_policy: Some("Retain".into()),
                storage_class_name: Some(storage_class.into()),
                claim_ref: Some(ObjectReference {
                    namespace: Some("database".into()),
                    name: Some(format!("capi-postgres-{ordinal}")),
                    ..Default::default()
                }),
                host_path: Some(HostPathVolumeSource {
                    path: format!(
                        "{}/volumes/cnpg/{ordinal}",
                        context.inputs.storage_container_path
                    ),
                    type_: Some("DirectoryOrCreate".into()),
                }),
                ..Default::default()
            }),
            ..Default::default()
        })
        .collect()
}

pub fn cnpg_cluster(
    context: &Context<'_>,
    storage_class: &str,
    postgres_image: &str,
) -> DynamicObject {
    let affinity = if context.spec.databases > context.spec.workers {
        "preferred"
    } else {
        "required"
    };
    context.object("postgresql.cnpg.io/v1", "Cluster", "capi-postgres", "database", "cnpg", json!({
        "instances":context.spec.databases, "imageName":postgres_image,
        "affinity":{"enablePodAntiAffinity":true,"podAntiAffinityType":affinity,"topologyKey":"kubernetes.io/hostname"},
        "bootstrap":{"initdb":{"database":"app","owner":"app"}},
        "storage":{"size":"1Gi","storageClass":storage_class},
        "resources":{"requests":{"cpu":"100m","memory":"256Mi"},"limits":{"cpu":"1","memory":"1Gi"}}
    }))
}

pub fn cnpg_objects(
    context: &Context<'_>,
    storage_class: &str,
    postgres_image: &str,
) -> Result<Vec<DynamicObject>, BuildError> {
    let mut objects = vec![to_dynamic(&database_namespace(context))?];
    for volume in persistent_volumes(context, storage_class) {
        objects.push(to_dynamic(&volume)?);
    }
    objects.push(cnpg_cluster(context, storage_class, postgres_image));
    Ok(objects)
}
